### ISSUE TITLE
jit tracing typecheck

### ISSUE THREAD

**Issue by nikhilmishra000** [pytorch/pytorch#16097]
## 🐛 Bug

Consider the following minimal example:
```
class InnerModule(nn.Module):
    
    def forward(self, x, y_dict):
        return x + y_dict['y']
     
class OuterModule(nn.Module):
    
    def __init__(self):
        super().__init__()
        self.inner = InnerModule()
        
    def forward(self, x, y):
        y_dict = dict(y=y)
        return self.inner(x, y_dict)
    
self = OuterModule()
x, y = torch.randn(3), torch.randn(3)

torch.jit.trace(self, (x, y)) # fails
```

This fails with the message:
```Auto nesting doesn't know how to process an input object of type dict. Accepted types: Tensors, or lists/tuples of them```

I can see why these type restrictions would exist for `OuterModule`, but it seems reasonable to me that the above behavior should be supported. The traced graph would be same, for instance, if `InnerModule` was changed to something like: 
```
class AlternativeInnerModule(nn.Module):
    
    def forward(self, x, y_tuple):
        return x + y_tuple[0]
```

I am using the 1.0 stable release installed from conda (ubuntu / python 3).

**Comment by eellison** [comment]
Thanks for raising the issue. We will look into this and get back to you. 

A work-around for now would be to put y_dict in the enclosing scope.

```
y_dict = dict(y=y)
class InnerModule(nn.Module):
    
    def forward(self, x, y_dict):
        return x + y_dict['y']
     
class OuterModule(nn.Module):
    
    def __init__(self):
        super().__init__()
        self.inner = InnerModule()
        
    def forward(self, x, y):
        y_dict = dict(y=y)
        return self.inner(x, y_dict)
```

**Comment by zdevito** [comment]
This should be classified as a bug. This is related to this apparently unused line in `_slow_forward`:

> input_vars = tuple(torch.autograd.function._iter_tensors(input))

**Comment by suo** [comment]
This is fixed in #16616
